lib/msal-common/src/telemetry/performance/PerformanceClient.ts

```diff
@@ -104,6 +122,102 @@ export abstract class PerformanceClient implements IPerformanceClient {
+     */
+    abstract setPreQueueTime(eventName: PerformanceEvents, correlationId?: string): void;
+
+    /**
+     * Gets map of pre-queue times by correlation Id
+     * 
+     * @param {PerformanceEvents} eventName 
+     * @param {string} correlationId 
+     * @returns {number}
+     */
+    getPreQueueTime(eventName: PerformanceEvents, correlationId: string): number | void {
+        const preQueueTimesByEvents = this.preQueueTimeByCorrelationId.get(correlationId);
+
+        if (!preQueueTimesByEvents) {
+            this.logger.trace(`PerformanceClient.getPreQueueTime: no pre-queue times found for correlationId: ${correlationId}, unable to add queue measurement`);
+            return;
+        } else if (!preQueueTimesByEvents.get(eventName)) {
+            this.logger.trace(`PerformanceClient.getPreQueueTime: no pre-queue time found for ${eventName}, unable to add queue measurement`);
+            return;
+        }
+
+        return preQueueTimesByEvents.get(eventName);
+    }
+
+    /**
+     * Calculates the difference between current time and time when function was queued.
+     * Note: It is possible to have 0 as the queue time if the current time and the queued time was the same.
+     * 
+     * @param {number} preQueueTime 
+     * @param {number} currentTime 
+     * @returns {number}
+     */
+    calculateQueuedTime(preQueueTime: number, currentTime: number): number {        
+        if (preQueueTime < 1) {
+            this.logger.trace(`PerformanceClient: preQueueTime should be a positive integer and not ${preQueueTime}`);
+            return 0;
+        }
+
+        if (currentTime < 1) {
+            this.logger.trace(`PerformanceClient: currentTime should be a positive integer and not ${currentTime}`);
+            return 0;
+        }
+
+        if (currentTime < preQueueTime) {
+            this.logger.trace("PerformanceClient: currentTime is less than preQueueTime, check how time is being retrieved");
+            return 0;
+        }
+
+        return currentTime-preQueueTime;
+    }
+
+    /**
+     * Adds queue measurement time to QueueMeasurements array for given correlation ID.
+     * 
+     * @param {PerformanceEvents} name 
+     * @param {?string} correlationId 
+     * @param {?number} time 
+     * @returns 
+     */
+    addQueueMeasurement(eventName: PerformanceEvents, correlationId?: string, queueTime?: number): void {
+
+        if (!correlationId) {
+            this.logger.trace(`PerformanceClient.addQueueMeasurement: correlationId not provided for ${eventName}, cannot add queue measurement`);
+            return;
+        }
+
+        if (queueTime === 0) {
+            // Possible for there to be no queue time after calculation
+            this.logger.trace(`PerformanceClient.addQueueMeasurement: queue time provided for ${eventName} is ${queueTime}`);
+        } else if (!queueTime) {
+            this.logger.trace(`PerformanceClient.addQueueMeasurement: no queue time provided for ${eventName}`);
```

Comment:
Maybe a future PBI (let's complete this PR) is when we have weirdness in our handling, like queue times are negative, or things are missing but expected we log a datapoint indicating something is off here with our timing and that data might be suspect.  Can you track an item for this?

Reply:
Got it. Work item created and tracked.